From adc7bebfd1e4dff1a2160fb587fc534f38463cba Mon Sep 17 00:00:00 2001
From: Sergio Lopez <slp@redhat.com>
Date: Thu, 28 May 2026 11:43:46 +0200
Subject: [PATCH 10/30] virtio/vsock: implement has_data for DGRAM

Signed-off-by: Sergio Lopez <slp@redhat.com>
---
diff --git a/include/linux/virtio_vsock.h b/include/linux/virtio_vsock.h
index e3d7afa29..34220effc 100644
--- a/include/linux/virtio_vsock.h
+++ b/include/linux/virtio_vsock.h
@@ -215,6 +215,7 @@ ssize_t
 virtio_transport_seqpacket_dequeue(struct vsock_sock *vsk,
 				   struct msghdr *msg,
 				   int flags);
+s64 virtio_transport_dgram_has_data(struct vsock_sock *vsk);
 s64 virtio_transport_stream_has_data(struct vsock_sock *vsk);
 s64 virtio_transport_stream_has_space(struct vsock_sock *vsk);
 u32 virtio_transport_seqpacket_has_data(struct vsock_sock *vsk);
diff --git a/include/net/af_vsock.h b/include/net/af_vsock.h
index 12b08459d..5f0e62555 100644
--- a/include/net/af_vsock.h
+++ b/include/net/af_vsock.h
@@ -76,6 +76,7 @@ struct vsock_sock {
 };
 
 s64 vsock_connectible_has_data(struct vsock_sock *vsk);
+s64 vsock_dgram_has_data(struct vsock_sock *vsk);
 s64 vsock_stream_has_data(struct vsock_sock *vsk);
 s64 vsock_stream_has_space(struct vsock_sock *vsk);
 struct sock *vsock_create_connected(struct sock *parent);
@@ -134,6 +135,7 @@ struct vsock_transport {
 	 * header.
 	 */
 	const size_t dgram_payload_offset;
+	s64 (*dgram_has_data)(struct vsock_sock *);
 
 	/* STREAM. */
 	/* TODO: stream_bind() */
diff --git a/net/vmw_vsock/af_vsock.c b/net/vmw_vsock/af_vsock.c
index 5d79fccc4..d534197ed 100644
--- a/net/vmw_vsock/af_vsock.c
+++ b/net/vmw_vsock/af_vsock.c
@@ -1007,6 +1007,15 @@ s64 vsock_stream_has_data(struct vsock_sock *vsk)
 }
 EXPORT_SYMBOL_GPL(vsock_stream_has_data);
 
+s64 vsock_dgram_has_data(struct vsock_sock *vsk)
+{
+	if (WARN_ON(!vsk->transport))
+		return 0;
+
+	return vsk->transport->dgram_has_data(vsk);
+}
+EXPORT_SYMBOL_GPL(vsock_dgram_has_data);
+
 s64 vsock_connectible_has_data(struct vsock_sock *vsk)
 {
 	struct sock *sk = sk_vsock(vsk);
@@ -1597,7 +1606,11 @@ static int vsock_do_ioctl(struct socket *sock, unsigned int cmd,
 			break;
 		}
 
-		n_bytes = vsock_stream_has_data(vsk);
+		if (sk->sk_type == SOCK_DGRAM)
+			n_bytes = vsock_dgram_has_data(vsk);
+		else
+			n_bytes = vsock_stream_has_data(vsk);
+
 		if (n_bytes < 0) {
 			ret = n_bytes;
 			break;
diff --git a/net/vmw_vsock/virtio_transport.c b/net/vmw_vsock/virtio_transport.c
index 48fd5a50e..3bf5f2484 100644
--- a/net/vmw_vsock/virtio_transport.c
+++ b/net/vmw_vsock/virtio_transport.c
@@ -559,6 +559,7 @@ static struct virtio_transport virtio_transport = {
 		.dgram_get_cid		  = virtio_transport_dgram_get_cid,
 		.dgram_get_port		  = virtio_transport_dgram_get_port,
 		.dgram_get_length	  = virtio_transport_dgram_get_length,
+		.dgram_has_data		  = virtio_transport_dgram_has_data,
 
 		.stream_dequeue           = virtio_transport_stream_dequeue,
 		.stream_enqueue           = virtio_transport_stream_enqueue,
diff --git a/net/vmw_vsock/virtio_transport_common.c b/net/vmw_vsock/virtio_transport_common.c
index 8798653b9..e2ce9f5a1 100644
--- a/net/vmw_vsock/virtio_transport_common.c
+++ b/net/vmw_vsock/virtio_transport_common.c
@@ -1012,6 +1012,21 @@ int virtio_transport_dgram_get_length(struct sk_buff *skb, size_t *len)
 }
 EXPORT_SYMBOL_GPL(virtio_transport_dgram_get_length);
 
+s64 virtio_transport_dgram_has_data(struct vsock_sock *vsk)
+{
+	struct sock *sk = sk_vsock(vsk);
+	struct sk_buff *skb;
+	s64 bytes = 0;
+
+	spin_lock_bh(&sk->sk_receive_queue.lock);
+	skb_queue_walk(&sk->sk_receive_queue, skb)
+		bytes += skb->len;
+	spin_unlock_bh(&sk->sk_receive_queue.lock);
+
+	return bytes;
+}
+EXPORT_SYMBOL_GPL(virtio_transport_dgram_has_data);
+
 s64 virtio_transport_stream_has_data(struct vsock_sock *vsk)
 {
 	struct virtio_vsock_sock *vvs = vsk->trans;
diff --git a/net/vmw_vsock/vsock_loopback.c b/net/vmw_vsock/vsock_loopback.c
index c50bebc88..d37e4f390 100644
--- a/net/vmw_vsock/vsock_loopback.c
+++ b/net/vmw_vsock/vsock_loopback.c
@@ -71,6 +71,7 @@ static struct virtio_transport loopback_transport = {
 		.dgram_get_cid		  = virtio_transport_dgram_get_cid,
 		.dgram_get_port		  = virtio_transport_dgram_get_port,
 		.dgram_get_length	  = virtio_transport_dgram_get_length,
+		.dgram_has_data		  = virtio_transport_dgram_has_data,
 
 		.stream_dequeue           = virtio_transport_stream_dequeue,
 		.stream_enqueue           = virtio_transport_stream_enqueue,
-- 
2.43.0
